### PR TITLE
Restore vectorization to full Poseidon rounds on Aarch64

### PR DESCRIPTION
**Pull request by nbgl**
Vectorized full round MDS for the new matrix. Minor changes to full S-boxes.

### CODE REVIEW

File: plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs [hunk 1]
```diff
@@ -2,37 +2,24 @@
...
-    1 << 16,
-];
-
-// The round constants to be applied by the second set of full rounds. These are just the usual round constants,
-// shifted by one round, with zeros shifted in.
+const EPSILON: u64 = 0xffffffff;
+
+// The round constants to be applied by the second set of full rounds. These are just the usual
+// round constants, shifted by one round, with zeros shifted in.
+/*
```
Comment (nbgl) [inline plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs:22]: This PR comments out a bunch of code in this file, which is now being compiled again. The commented code will be used in the partial rounds.

File: plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs [hunk 2]
```diff
@@ -52,9 +40,12 @@ const FINAL_ROUND_CONSTANTS: [u64; WIDTH * HALF_N_FULL_ROUNDS] = make_final_roun
 const fn check_mds_matrix() -> bool {
     // Can't == two arrays in a const_assert! (:
     let mut i = 0;
-    let wanted_matrix_exps = [0, 0, 1, 0, 3, 5, 1, 8, 12, 3, 16, 10];
+    let wanted_matrix_circ = [17, 15, 41, 16, 2, 28, 13, 13, 39, 18, 34, 20];
+    let wanted_matrix_diag = [8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
```
Comment (unzvfu) [inline plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs:44]: I feel you might have explained this before, but why is the circulant matrix above the (almost) reverse of the matrix I originally gave: `[17, 20, 34, 18, 39, 13, 13, 28, 2, 16, 41, 15]`?

Comment (unzvfu) [reply plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs:44]: Oh,  I guess the transpose has been taken somewhere.

File: plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs [hunk 3]
```diff
@@ -133,7 +95,16 @@ unsafe fn add_with_wraparound(a: u64, b: u64) -> u64 {
 /// Subtraction of a and (b >> 32) modulo ORDER accounting for wraparound.
 #[inline(always)]
 unsafe fn sub_with_wraparound_lsr32(a: u64, b: u64) -> u64 {
-    let b_hi = b >> 32;
+    let mut b_hi = b >> 32;
+    // Make sure that LLVM emits two separate instructions for the shift and the subtraction. This
+    // reduces pressure on the execution units with access to the flags, as they are no longer
+    // responsible for othe shift. The hack is to inserts a fake computation between the two
```
Comment (unzvfu) [inline plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs:101]: *other
*insert

**Review by unzvfu — APPROVED** [review]
This looks good to me, though I don't have an M1 to test it.

File: plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs (hunk 3 above)
Comment (nbgl) [reply plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs:101]: jakub needs too profread

File: plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs (hunk 2 above)
Comment (nbgl) [reply plonky2/src/hash/arch/aarch64/poseidon_goldilocks_neon.rs:44]: Yeah, the difference is whether you're storing the first row or the first column of a circulant matrix. The generic Poseidon code stores the first row. But to do convolution in Fourier space, you actually need to take the FFT of the first column. So it's the matrix you gave me, I just define it by storing a different thing.

**Pull request merged**